### PR TITLE
Limit max connection pool size

### PR DESCRIPTION
**Pull request by jshore1296**

### CODE REVIEW

File: memcache/memcache.go [hunk 1]
```diff
@@ -68,15 +68,21 @@ var (
 
 	// ErrClusterConfigMiss means that GetConfig failed as cluster config was not present
 	ErrClusterConfigMiss = errors.New("memcache: cluster config miss")
+
+	ErrConnectionPoolTimeout = errors.New("memcache: connection pool timeout")
```
Comment (sgp) [inline memcache/memcache.go:74]: Needs godoc

File: memcache/memcache.go [hunk 2]
```diff
@@ -68,15 +68,21 @@ var (
...
 )
 
 const (
 	// DefaultTimeout is the default socket read/write timeout.
 	DefaultTimeout = 100 * time.Millisecond
 
+	DefaultPoolTimeout = 2 * DefaultTimeout
```
Comment (sgp) [inline memcache/memcache.go:82]: Needs godoc

File: memcache/memcache.go [hunk 3]
```diff
@@ -184,6 +190,10 @@ type Client struct {
 	// be set to a number higher than your peak parallel requests.
 	MaxIdleConns int
 
+	PoolTimeout    time.Duration
```
Comment (sgp) [inline memcache/memcache.go]: Needs godoc

File: memcache/memcache.go [hunk 4]
```diff
@@ -68,15 +68,21 @@ var (
...
 	// DefaultTimeout is the default socket read/write timeout.
 	DefaultTimeout = 100 * time.Millisecond
 
+	DefaultPoolTimeout = 2 * DefaultTimeout
+
 	// DefaultMaxIdleConns is the default maximum number of idle connections
 	// kept for any single address.
 	DefaultMaxIdleConns = 2
+
+	DefaultMaxActiveConns = 10
```
Comment (sgp) [inline memcache/memcache.go:89]: Needs godoc

File: memcache/memcache.go [hunk 5]
```diff
@@ -233,12 +243,13 @@ func (cn *conn) extendDeadline() {
...
 // cache miss).  The purpose is to not recycle TCP connections that
 // are bad.
-func (cn *conn) condRelease(err *error) {
+func (cn *conn) condRelease(err *error, lock chan struct{}) {
 	if *err == nil || resumableError(*err) {
 		cn.release()
 	} else {
 		cn.nc.Close()
 	}
+	<-lock
```
Comment (sgp) [inline memcache/memcache.go]: should this be in a `defer`?

File: memcache/memcache.go [hunk 6]
```diff
@@ -313,14 +348,52 @@ func (c *Client) dial(addr net.Addr) (net.Conn, error) {
...
+		}
+		c.lk.Unlock()
+	}
+	if c.activeConnLock[addr.String()] == nil {
+		c.lk.Lock()
+		if c.activeConnLock[addr.String()] == nil {
+			c.activeConnLock[addr.String()] = make(chan struct{}, c.maxConns())
+		}
+		c.lk.Unlock()
+	}
```
Comment (sgp) [inline memcache/memcache.go:390]: I'd take this entire codeblock and stick it in a method off of `*Client` just for readability.

File: memcache/memcache_discovery_test.go [hunk 7]
```diff
@@ -197,6 +197,7 @@ func startMemcachedServer(t *testing.T, port int, ready chan<- bool, exit <-chan
 	cmd := exec.Command("memcached", "-p", strconv.Itoa(port))
 	t.Logf("starting memcached server with command : %v", cmd)
 	if err := cmd.Start(); err != nil {
+		t.Log(err)
```
Comment (sgp) [inline memcache/memcache_discovery_test.go:200]: Kill this with :fire:

**Review by sgp — CHANGES_REQUESTED** [review]

File: memcache/memcache.go (hunk 5 above)
Comment (jshore1296) [reply memcache/memcache.go]: yeah, I guess.  easier than freaking out about panics on release or Close()

File: memcache/memcache_discovery_test.go (hunk 7 above)
Comment (jshore1296) [reply memcache/memcache_discovery_test.go:200]: This is... in a test?  And it's actually quite helpful to know why the test failed here.

**Review by sgp — APPROVED** [review]

**Review by sgp — APPROVED** [review]

**Pull request merged**